### ISSUE TITLE
instabilities

### ISSUE THREAD

**Issue by folkertvanheusden** [davidmoreno/rtpmidid#55]
I notice that quite often connections "go away". I'm trying to figure this out.
For this, I collect the odd things I notice (errors, etc.).

```
Jul 31 17:56:30 lappiemctopface rtpmidid[135084]: #033[0;34m[2021-07-31T15:56:25Z] [rtpmidid.cpp:275]#033[0m Found client to delete: alsa port 3. Deletes all known addreses.
Jul 31 17:56:30 lappiemctopface rtpmidid[135084]: #033[0;34m[2021-07-31T15:56:25Z] [rtpmidid.cpp:574]#033[0m Removing peer from known peers list. Port 3
Jul 31 17:57:53 lappiemctopface rtpmidid[135084]: #033[0;34m[2021-07-31T15:57:48Z] [rtpclient.cpp:156]#033[0m MIDI PORT at port 39352
Jul 31 17:57:53 lappiemctopface rtpmidid[135084]: #033[0;34m[2021-07-31T15:57:48Z] [rtpclient.cpp:182]#033[0m Connecting control port 39352 to ESP-76E44C.dhcp.nurd.space:123
Jul 31 20:24:43 lappiemctopface rtpmidid[162733]: #033[0;34m[2021-07-31T18:24:27Z] [rtpclient.cpp:156]#033[0m MIDI PORT at port 38608
Jul 31 20:24:43 lappiemctopface rtpmidid[162733]: #033[0;34m[2021-07-31T18:24:27Z] [rtpclient.cpp:182]#033[0m Connecting control port 38608 to 192.168.122.178:5004
Jul 31 20:24:43 lappiemctopface rtpmidid[162733]: #033[0;34m[2021-07-31T18:24:27Z] [rtpclient.cpp:188]#033[0m Connecting midi port 38609 to 192.168.122.178:5005
Jul 31 20:32:34 lappiemctopface rtpmidid[162733]: #033[0;31m[2021-07-31T18:32:34Z] [rtpmidid.cpp:479]#033[0m There is no peer but I received an event! This situation should NEVER happen. File a bug. Port 3
Jul 31 20:32:36 lappiemctopface rtpmidid[162733]: message repeated 6 times: [ #033[0;31m[2021-07-31T18:32:34Z] [rtpmidid.cpp:479]#033[0m There is no peer but I received an event! This situation should NEVER happen. File a bug. Port 3]
Jul 31 20:32:36 lappiemctopface rtpmidid[162733]: #033[0;31m[2021-07-31T18:32:35Z] [rtpmidid.cpp:479]#033[0m There is no peer but I received an event! This situation should NEVER happen. File a bug. Port 3
Jul 31 20:32:36 lappiemctopface rtpmidid[162733]: message repeated 10 times: [ #033[0;31m[2021-07-31T18:32:35Z] [rtpmidid.cpp:479]#033[0m There is no peer but I received an event! This situation should NEVER happen. File a bug. Port 3]
Jul 31 20:32:36 lappiemctopface rtpmidid[162733]: #033[0;31m[2021-07-31T18:32:36Z] [rtpmidid.cpp:479]#033[0m There is no peer but I received an event! This situation should NEVER happen. File a bug. Port 3
Jul 31 20:32:36 lappiemctopface rtpmidid[162733]: message repeated 13 times: [ #033[0;31m[2021-07-31T18:32:36Z] [rtpmidid.cpp:479]#033[0m There is no peer but I received an event! This situation should NEVER happen. File a bug. Port 3]
```

**Comment by folkertvanheusden** [comment]
[2021-08-02T20:21:40Z] [poller.cpp:200] Caught exception at poller: Try to access end of buffer at 1

That's when talking to a https://github.com/lathoub/Arduino-AppleMIDI-Library

**Comment by folkertvanheusden** [comment]
This is an example of a disappearing service.
I had a '' and a ''. As soon as I connected them (in a patchbay program), the following scrolled by in the rtpmidid output:

```
[2021-08-02T21:20:55Z] [mdns_rtpmidi.cpp:235] Discovered service 'AppleMIDI-ESP8266' of type '_apple-midi._udp' in domain 'local'
[2021-08-02T21:20:55Z] [rtpmidid.cpp:243] New alsa port: 3, connects to host: 192.168.65.156, port: 5004, name: AppleMIDI-ESP8266
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:58] Service '/Client39/EntryGroup1' successfully established
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:284] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:284] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:284] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:284] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:231] Received own announcement
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:231] Received own announcement
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:231] Received own announcement
[2021-08-02T21:20:56Z] [mdns_rtpmidi.cpp:231] Received own announcement
[2021-08-02T21:21:53Z] [aseq.cpp:123] New ALSA connection from port ESI MIDIMATE eX-ESI MIDIMATE eX MIDI 1 (20:0)
[2021-08-02T21:21:53Z] [rtpmidid.cpp:249] Callback on subscribe at rtpmidid: ESI MIDIMATE eX-ESI MIDIMATE eX MIDI 1
[2021-08-02T21:21:53Z] [rtpclient.cpp:77] Try connect to service at 192.168.65.156:5004
[2021-08-02T21:21:53Z] [rtpclient.cpp:99] Try connect to resolved name: 192.168.65.156:5004
[2021-08-02T21:21:53Z] [rtpclient.cpp:118] Connected to resolved name: 192.168.65.156:5004
[2021-08-02T21:21:53Z] [rtpclient.cpp:126] Control port, local: 32769, remote at 192.168.65.156:5004
[2021-08-02T21:21:53Z] [rtpclient.cpp:156] MIDI PORT at port 32769
[2021-08-02T21:21:53Z] [rtpclient.cpp:182] Connecting control port 32769 to 192.168.65.156:5004
[2021-08-02T21:21:58Z] [rtpmidid.cpp:313] Disconnect signal: 4
[2021-08-02T21:21:58Z] [rtpmidid.cpp:341] Timeout. Not trying again.
[2021-08-02T21:21:58Z] [rtpmidid.cpp:574] Removing peer from known peers list. Port 3
[2021-08-02T21:21:58Z] [rtppeer.cpp:50] ~rtppeer 'lappiemctopface/ESI MIDIMATE eX-ESI MIDIMATE eX MIDI 1' (local) <-> '' (remote)
```

"ESI MIDIMATE eX-ESI MIDIMATE eX MIDI 1" is a midi-to-usb-interface on my laptop, 192.168.65.156:5004 is the 'AppleMIDI-ESP8266'.

Now what I find puzzling is why this is printed:

```
[2021-08-02T21:21:53Z] [rtpclient.cpp:126] Control port, local: 32769, remote at 192.168.65.156:5004
[2021-08-02T21:21:53Z] [rtpclient.cpp:156] MIDI PORT at port 32769
```

Here the control port and the midi-port are the same port?
This is with this version by the way:

```
commit adbe2c53fba5a5c46b1f3f67496b88c6f7cc81bf (HEAD)
Author: David Moreno <dmoreno@coralbits.com>
Date:   Fri Jul 23 19:51:32 2021 +0200
```
so without any modifications by me.

In lib/rtpclient.cpp, there's
```
void rtpclient::connect_to(const std::string &address,
                           const std::string &port) {
```
and at line 154 you see:
```
    ::getsockname(control_socket, (struct sockaddr *)&servaddr, &len);
    auto midi_port = htons(servaddr.sin6_port);
    DEBUG("MIDI PORT at port {}", midi_port);
```

Shouldn't that be "::getsockname(midi_socket, (struct sockaddr *)&servaddr, &len);" ?

**Comment by folkertvanheusden** [comment]
Observation: the problem only occurs over wifi here. I thought maybe it was a timing problem, but I succesfully connected over a vpn to my home via rtpmidi and it ran for quite a while with that.
From what I saw in the source code, it is avahi telling rtpmidi that the other end is gone.

**Comment by folkertvanheusden** [comment]
Also strange: sometimes connecting some source to a local fluidsynth works fine ONLY if I also connect it to an other service. See screendumps to see what I mean.
![silence](https://user-images.githubusercontent.com/85709915/128644271-a9a41e82-bac1-4e1f-8073-f2de4fe284f0.png)
![works-fine](https://user-images.githubusercontent.com/85709915/128644281-a171c70d-47fa-4312-8cf0-b0988b507c79.png)

**Comment by davidmoreno** [comment]
I also detected the other day something similar, but thought it was related to disconnection. My case was connecting to two alsa ports, and disconnecting one of them. Then no data is sent until another port is connected. Do you think that could be your case as well?

**Comment by davidmoreno** [comment]
> update: there is **NO** difference between ethernet and wifi. It still may be a timing problem, not perse a too long latency (over vpn worked fine for example (aprox. 100ms latency)).
> 
> From what I saw in the source code, it is avahi telling rtpmidi that the other end is gone.
> 
> How does it handle packet loss? At home I have about 2.1% packetloss (wifi or ethernet(-over-power), both not 100% reliable - strangely enough 0% packet loss over vpn).

On VPN quite probably the VPN (wireguard) itself is doing some packet resend. Is the VPN directly to the rtpmidid computer?

There is no rtpmidi resend support, for that packet loss, whats needed is the journaling support, which is not finished yet. Are you just using a LAN for rtpmidi? or souring though internet or something? 2.1% packet loss is a lot for a LAN IMHO. For me I have 0% (or not quite sure how to check it, 0 packets drops at ifconfig)

**Comment by folkertvanheusden** [comment]
> I also detected the other day something similar, but thought it was related to disconnection. My case was connecting to two alsa ports, and disconnecting one of them. Then no data is sent until another port is connected. Do you think that could be your case as well?

Yes. And also that it did not start to work before I connect two.

**Comment by folkertvanheusden** [comment]
> On VPN quite probably the VPN (wireguard) itself is doing some packet resend. Is the VPN directly to the rtpmidid computer?

I use tinc; its endpoint is on a virtual machine. Other virtual machines (on the same server) run rtpmidi, usually with my 'metronoom' test-program.


> There is no rtpmidi resend support, for that packet loss, whats needed is the journaling support, which is not finished yet. Are you just using a LAN for rtpmidi? or souring though internet or something? 2.1% packet loss is a lot for a LAN IMHO. For me I have 0% (or not quite sure how to check it, 0 packets drops at ifconfig)

Well it depends. I tested it again and this time I've got 0% packet loss on wifi. The ethernet is over power in my living room, and that's somewhat comparable to ethernet over radio (wifi) and thus prone to packet loss.
But also without packetloss the problems occure.

**Comment by folkertvanheusden** [comment]
> > I also detected the other day something similar, but thought it was related to disconnection. My case was connecting to two alsa ports, and disconnecting one of them. Then no data is sent until another port is connected. Do you think that could be your case as well?
> 
> Yes. And also that it did not start to work before I connect two.

Oh and note: in this example, 'neo' and 'arduino' are directly on the same switch (no vpn, wifi or eop in between).
Unfortunately I can't remember if 'lappiemctopface' was on wifi or lan at that time (maybe add a hint for which interface one is connected in the name?).

**Comment by folkertvanheusden** [comment]
Please check this logging: it appears that suddenly rtpmidid decides there's a timeout and disconnects.

[river_rtpmidi_fail.txt](https://github.com/davidmoreno/rtpmidid/files/6974536/river_rtpmidi_fail.txt)

**Comment by folkertvanheusden** [comment]
I did a quick rtpmidi implementation to debug the problems. Now especially at the beginning but also during the session I often receive 'ICMP Destination unreachable' messages. I think that's a bit strange as rtpmidid is supposed to always be listening on those 2 ports?

```
08:53:03.553474 IP 192.168.64.100 > 192.168.65.205: ICMP 192.168.64.100 udp port 36775 unreachable, length 48
08:53:03.599573 IP 192.168.64.100 > 192.168.65.205: ICMP 192.168.64.100 udp port 36775 unreachable, length 48
```
Network trace: [rtpmidi.pcap.zip](https://github.com/davidmoreno/rtpmidid/files/6980461/rtpmidi.pcap.zip)

**Comment by davidmoreno** [comment]
Its strange that it happens only on port 5004, but it was properly used before.. On a moment I thouhg t it it could be caused byt the network packet losses you were experiencing, but the ICMP message is sent by the proper OS.. so it should be ok.

Maybe some timing when creating the ports? I think we create one first and then the second after the first is connected. Maybe we can create them both at the same time so its ready for later? But it does not make sense as some packets seem to arrive and others not.

anyway t seems to be totally related to the journaling, as its on feedback packets that it gets the errors.

I would try to create some test program that is able to reproduce the problem reliablily, so we can fix it completely.

**Comment by folkertvanheusden** [comment]
I'll put my test-program on-line: https://github.com/folkertvanheusden/frtpm

It hardcoded listens on port 5004/5005 and creates an ALSA port on which everything is outputted.

```
rtmidid-cli create rftpm ...IP-ADDRESS... 5004
```

should work.

**Comment by folkertvanheusden** [comment]
In void rtppeer::parse_midi(io_bytes_reader &buffer) there's this code:

```
  buffer.read_uint8(); // Ignore RTP header flags (Byte 0)
  auto rtpmidi_id = buffer.read_uint8();
  if (rtpmidi_id != 0x61) { // next Byte: Payload type
```

Shouldn't you mask off the 7th bit? Because that contains 'M', indicating if there's MIDI data and that can be set as well I believe:

```
| M               |   8   | Has MIDI data.                                                                                                           |
| PT              | 9-15  | Payload type. Always 0x61. MIDI.                                                                                         |
```
(from docs/RFC6295_notes.md).

From 'RFC 6295':
```
"For native streams, the M bit MUST be set to 1 if the MIDI
   command section has a non-zero LEN field and MUST be set to 0
   otherwise."
```

**Comment by folkertvanheusden** [comment]
I bought a 'Doremidi MR-1' appliance which interfaces serial-midi (din-5 plugs) with rtpmidi/apple-midi.
Also this device (something embedded) appears and then disappears. This disappearing happens a few seconds after I connect some other midi-device to it (under linux in patchage).

**Comment by folkertvanheusden** [comment]
I have a '[DOREMiDi](https://www.ebay.nl/itm/265215356404)' which is a converter between midi and rtpmidi/apple-midi.
It is an appliance running on some kind of microcontroller (boots instantly).
It connects fine with [windows](http://www.tobias-erichsen.de/software/rtpmidi.html) but with rtpmidi I've got it to connect sofar only once. I started a post about this earlier and then deleted but that was too soon :-)

```[0] [mdns_rtpmidi.cpp:265] Discovered service 'DOREMiDi-MIDI' of type '_apple-midi._udp' in domain 'local'
[3] [rtpmidid.cpp:243] New alsa port: 8, connects to host: 192.168.65.184, port: 5006, name: DOREMiDi-MIDI
[0] [mdns_rtpmidi.cpp:265] Discovered service 'belle' of type '_apple-midi._udp' in domain 'local'
[0] [mdns_rtpmidi.cpp:265] Discovered service 'metronoom' of type '_apple-midi._udp' in domain 'local'
[0] [mdns_rtpmidi.cpp:265] Discovered service 'neo' of type '_apple-midi._udp' in domain 'local'
[0] [mdns_rtpmidi.cpp:265] Discovered service 'cameras4' of type '_apple-midi._udp' in domain 'local'
[0] [mdns_rtpmidi.cpp:265] Discovered service 'metronoom_river' of type '_apple-midi._udp' in domain 'local'
[0] [mdns_rtpmidi.cpp:265] Discovered service 'river' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:64] Service '/Client33/EntryGroup1' successfully established
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[3] [mdns_rtpmidi.cpp:316] (Browser) NEW: service 'lappiemctopface' of type '_apple-midi._udp' in domain 'local'
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[0] [mdns_rtpmidi.cpp:261] Received own announcement
[3] [aseq.cpp:128] New ALSA connection from port Midi Through-Midi Through Port-0 (14:0)
[0] [rtpmidid.cpp:249] Callback on subscribe at rtpmidid: Midi Through-Midi Through Port-0
[0] [rtpclient.cpp:80] Try connect to service at 192.168.65.184:5006
[0] [rtpclient.cpp:104] Try connect to resolved name: 192.168.65.184:5006
[0] [rtpclient.cpp:124] Connected to resolved name: 192.168.65.184:5006
[0] [rtpclient.cpp:132] Control port, local: 33165, remote at 192.168.65.184:5006
[0] [rtpclient.cpp:162] MIDI PORT at port 33166
[0] [rtpclient.cpp:188] Connecting midi port 33166 to 192.168.65.184:5007
[0] [rtpmidid.cpp:313] Disconnect signal: 4
[1] [rtpmidid.cpp:341] Timeout (during setup). Not trying again.
[0] [rtpmidid.cpp:595] Removing peer from known peers list. Port 8
[0] [rtppeer.cpp:52] ~rtppeer 'lappiemctopface/Midi Through-Midi Through Port-0' (local) <-> '' (remote)```

192.168.65.184 is the IP address of the doremidi appliance.

**Comment by folkertvanheusden** [comment]
The arduino AND the DOREMiDi (according to tcpdump) just don't respond to the invitation request. Requests look sane (also according to wireshark).

**Comment by folkertvanheusden** [comment]
Sometimes when rtpmidid decides to remove an rtp-midi-source, it goes into a loop doing only this:

```
recvfrom(18, 0x7ffe4f8c2800, 1500, MSG_DONTWAIT, 0x7ffe4f8c27c0, [28]) = -1 EBADF (Bad file descriptor)
epoll_wait(3, [{events=EPOLLIN, data={u32=10, u64=10}}], 10, -1) = 1
```

```
#0  0x00007f57ac4a32a0 in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#1  0x00007f57ac4a49f2 in _Unwind_RaiseException () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#2  0x00007f57ac55b7fc in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x000056392b9486e7 in rtpmidid::rtpclient::data_ready (this=0x56392c81f120, port=rtpmidid::rtppeer::MIDI_PORT)
    at /home/folkert/t/rtpmidid/lib/rtpclient.cpp:283
#4  0x000056392b946ed7 in operator() (__closure=0x56392c851858) at /home/folkert/t/rtpmidid/lib/rtpclient.cpp:163
#5  0x000056392b94a544 in std::__invoke_impl<void, rtpmidid::rtpclient::connect_to(const string&, const string&)::<lambda(int)>&, int>(std::__invoke_other, struct {...} &) (__f=...) at /usr/include/c++/10/bits/invoke.h:60
#6  0x000056392b9499a6 in std::__invoke_r<void, rtpmidid::rtpclient::connect_to(const string&, const string&)::<lambda(int)>&, int>(struct {...} &) (
    __fn=...) at /usr/include/c++/10/bits/invoke.h:110
#7  0x000056392b948f8e in std::_Function_handler<void(int), rtpmidid::rtpclient::connect_to(const string&, const string&)::<lambda(int)> >::_M_invoke(const std::_Any_data &, int &&) (__functor=..., __args#0=@0x7ffe4f8c2ec4: 10) at /usr/include/c++/10/bits/std_function.h:291
#8  0x000056392b963900 in std::function<void (int)>::operator()(int) const (this=0x56392c851858, __args#0=10)
    at /usr/include/c++/10/bits/std_function.h:622
#9  0x000056392b961107 in rtpmidid::poller_t::wait (this=0x56392b9b91a8 <rtpmidid::poller>) at /home/folkert/t/rtpmidid/lib/poller.cpp:202
#10 0x000056392b8e705f in main (argc=1, argv=0x7ffe4f8c3468) at /home/folkert/t/rtpmidid/src/main.cpp:65
```

**Comment by folkertvanheusden** [comment]
```
Aug 20 23:06:40 lappiemctopface rtpmidid[33230]: [poller.cpp:204] Caught exception at poller: Filedescriptor 30 for �#003���U, forgetting fd 24.
```

Looks as if thread A removes something while B is still referring to it.

**Comment by davidmoreno** [comment]
I don't really know if finally fixed in latest version, as I dont have access to DOREMIDI. If still broken, please open a new issue.

Thanks for the report and all the information.
